File: backend/src/main/java/com/technicalescaperoom/backend/controller/player/PlayerSessionController.java
package com.technicalescaperoom.backend.controller.player;

import com.technicalescaperoom.backend.config.security.PlayerPrincipal;
import com.technicalescaperoom.backend.dto.player.PlayerLoginRequest;
import com.technicalescaperoom.backend.dto.player.PlayerResponseDto;
import com.technicalescaperoom.backend.service.PlayerSessionService;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/player")
@RequiredArgsConstructor
public class PlayerSessionController {

    private final PlayerSessionService playerSessionService;

    @PostMapping("/login")
    public ResponseEntity<PlayerResponseDto> login(
            @Valid @RequestBody PlayerLoginRequest request,
            HttpServletRequest httpRequest,
            HttpServletResponse response
    ) {
        PlayerResponseDto playerResponse = playerSessionService.login(request, httpRequest, response);
        return ResponseEntity.ok(playerResponse);
    }

    @GetMapping("/me")
    public ResponseEntity<PlayerResponseDto> getCurrentPlayer(@AuthenticationPrincipal PlayerPrincipal principal) {
        PlayerResponseDto playerResponse = playerSessionService.getCurrentPlayer(principal);
        return ResponseEntity.ok(playerResponse);
    }

    @GetMapping("/lobby")
    public ResponseEntity<PlayerResponseDto> getLobbyState(@AuthenticationPrincipal PlayerPrincipal principal) {
        PlayerResponseDto lobbyState = playerSessionService.getLobbyState(principal);
        return ResponseEntity.ok(lobbyState);
    }

    @PostMapping("/ready")
    public ResponseEntity<PlayerResponseDto> setReady(
            @AuthenticationPrincipal PlayerPrincipal principal,
            @RequestBody(required = false) Map<String, Boolean> body
    ) {
        boolean isReady = body == null || body.get("ready") == null || Boolean.TRUE.equals(body.get("ready"));
        PlayerResponseDto updated = playerSessionService.setPlayerReady(principal, isReady);
        return ResponseEntity.ok(updated);
    }

    @PostMapping("/event/start")
    public ResponseEntity<PlayerResponseDto> startEvent(@AuthenticationPrincipal PlayerPrincipal principal) {
        PlayerResponseDto response = playerSessionService.startTeamEvent(principal);
        return ResponseEntity.ok(response);
    }

    @PostMapping("/logout")
    public ResponseEntity<Map<String, String>> logout(
            @AuthenticationPrincipal PlayerPrincipal principal,
            HttpServletResponse response
    ) {
        playerSessionService.logout(principal, response);
        return ResponseEntity.ok(Map.of("message", "Logged out successfully"));
    }
}

File: backend/controllers/player/playerSessionController.js
const playerSessionService = require('../../services/playerSessionService');

class PlayerSessionController {
  async login(req, res, next) {
    try {
      const { teamCode, playerNumber } = req.body;
      if (!teamCode || playerNumber == null) {
        return res.status(400).json({
          status: 400,
          error: 'Bad Request',
          code: 'BAD_REQUEST',
          message: 'teamCode and playerNumber are required',
          timestamp: new Date().toISOString()
        });
      }

      const response = await playerSessionService.login({ teamCode, playerNumber }, req, res);
      return res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async getCurrentPlayer(req, res, next) {
    try {
      const response = await playerSessionService.getCurrentPlayer(req.playerPrincipal);
      return res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async getLobbyState(req, res, next) {
    try {
      const response = await playerSessionService.getLobbyState(req.playerPrincipal);
      return res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async setReady(req, res, next) {
    try {
      const isReady = req.body == null || req.body.ready == null || req.body.ready === true;
      const response = await playerSessionService.setPlayerReady(req.playerPrincipal, isReady);
      return res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async startEvent(req, res, next) {
    try {
      const response = await playerSessionService.startTeamEvent(req.playerPrincipal);
      return res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async logout(req, res, next) {
    try {
      await playerSessionService.logout(req.playerPrincipal, res);
      return res.json({ message: 'Logged out successfully' });
    } catch (err) {
      next(err);
    }
  }
}

module.exports = new PlayerSessionController();
